### PR TITLE
Add tests for preprocess and backtrack workflow

### PR DESCRIPTION
**Pull request by geek-yang**
In this PR we add integration tests for the preprocess and backtrack workflow. Tests are performed by invoking the cli commands. Here is a brief list of changes:

- Add small (cropped) ERA5 data sets for testing
- Script for the preparation of testing data and add a readme
- Add preprocess & backtracks outputs data for verification
- Test preprocess workflow and verify the output
- Test backtracks workflow and verify the output

This PR refs #261. It addresses the first two points.

### CODE REVIEW

File: src/wam2layers/analysis/checks.py [hunk 1]
```diff
@@ -10,7 +10,7 @@ def _warning_on_one_line(message, category, filename, lineno, file=None, line=No
 
     https://stackoverflow.com/a/26433913
     """
-    short_path = Path(filename).relative_to(Path(filename).parents[2])
+    short_path = Path(filename).relative_to(Path(filename).parents[0])
```
Comment (geek-yang) [inline src/wam2layers/analysis/checks.py:13]: Change made here to fix the following error:

```py
❯ pytest tests/test_workflow.py
================================================= test session starts ==================================================
platform linux -- Python 3.8.10, pytest-7.4.2, pluggy-1.3.0
rootdir: /home/yangliu/WAM2layers/WAM2layers
collected 2 items

tests/test_workflow.py .
INTERNALERROR> Traceback (most recent call last):
INTERNALERROR>   File "/home/yangliu/venv/wam2layers/lib/python3.8/site-packages/_pytest/main.py", line 271, in wrap_session
INTERNALERROR>     session.exitstatus = doit(config, session) or 0
INTERNALERROR>   File "/home/yangliu/venv/wam2layers/lib/python3.8/site-packages/_pytest/main.py", line 325, in _main
INTERNALERROR>     config.hook.pytest_runtestloop(session=session)
INTERNALERROR>   File "/home/yangliu/venv/wam2layers/lib/python3.8/site-packages/pluggy/_hooks.py", line 493, in __call__
INTERNALERROR>     return self._hookexec(self.name, self._hookimpls, kwargs, firstresult)
INTERNALERROR>   File "/home/yangliu/venv/wam2layers/lib/python3.8/site-packages/pluggy/_manager.py", line 115, in _hookexec
INTERNALERROR>     return self._inner_hookexec(hook_name, methods, kwargs, firstresult)
INTERNALERROR>   File "/home/yangliu/venv/wam2layers/lib/python3.8/site-packages/pluggy/_callers.py", line 152, in _multicall
INTERNALERROR>     return outcome.get_result()
INTERNALERROR>   File "/home/yangliu/venv/wam2layers/lib/python3.8/site-packages/pluggy/_result.py", line 114, in get_result
INTERNALERROR>     raise exc.with_traceback(exc.__traceback__)
INTERNALERROR>   File "/home/yangliu/venv/wam2layers/lib/python3.8/site-packages/pluggy/_callers.py", line 77, in _multicall
INTERNALERROR>     res = hook_impl.function(*args)
INTERNALERROR>   File "/home/yangliu/venv/wam2layers/lib/python3.8/site-packages/_pytest/main.py", line 350, in pytest_runtestloop
INTERNALERROR>     item.config.hook.pytest_runtest_protocol(item=item, nextitem=nextitem)
INTERNALERROR>   File "/home/yangliu/venv/wam2layers/lib/python3.8/site-packages/pluggy/_hooks.py", line 493, in __call__
INTERNALERROR>     return self._hookexec(self.name, self._hookimpls, kwargs, firstresult)
INTERNALERROR>   File "/home/yangliu/venv/wam2layers/lib/python3.8/site-packages/pluggy/_manager.py", line 115, in _hookexec
INTERNALERROR>     return self._inner_hookexec(hook_name, methods, kwargs, firstresult)
INTERNALERROR>   File "/home/yangliu/venv/wam2layers/lib/python3.8/site-packages/pluggy/_callers.py", line 130, in _multicall
INTERNALERROR>     teardown[0].send(outcome)
INTERNALERROR>   File "/home/yangliu/venv/wam2layers/lib/python3.8/site-packages/_pytest/warnings.py", line 109, in pytest_runtest_protocol
INTERNALERROR>     yield
INTERNALERROR>   File "/usr/lib/python3.8/contextlib.py", line 120, in __exit__
INTERNALERROR>     next(self.gen)
INTERNALERROR>   File "/home/yangliu/venv/wam2layers/lib/python3.8/site-packages/_pytest/warnings.py", line 64, in catch_warnings_for_item
INTERNALERROR>     ihook.pytest_warning_recorded.call_historic(
INTERNALERROR>   File "/home/yangliu/venv/wam2layers/lib/python3.8/site-packages/pluggy/_hooks.py", line 514, in call_historic
INTERNALERROR>     res = self._hookexec(self.name, self._hookimpls, kwargs, False)
INTERNALERROR>   File "/home/yangliu/venv/wam2layers/lib/python3.8/site-packages/pluggy/_manager.py", line 115, in _hookexec
INTERNALERROR>     return self._inner_hookexec(hook_name, methods, kwargs, firstresult)
INTERNALERROR>   File "/home/yangliu/venv/wam2layers/lib/python3.8/site-packages/pluggy/_callers.py", line 113, in _multicall
INTERNALERROR>     raise exception.with_traceback(exception.__traceback__)
INTERNALERROR>   File "/home/yangliu/venv/wam2layers/lib/python3.8/site-packages/pluggy/_callers.py", line 77, in _multicall
INTERNALERROR>     res = hook_impl.function(*args)
INTERNALERROR>   File "/home/yangliu/venv/wam2layers/lib/python3.8/site-packages/_pytest/terminal.py", line 538, in pytest_warning_recorded
INTERNALERROR>     message = warning_record_to_str(warning_message)
INTERNALERROR>   File "/home/yangliu/venv/wam2layers/lib/python3.8/site-packages/_pytest/warnings.py", line 77, in warning_record_to_str
INTERNALERROR>     msg = warnings.formatwarning(
INTERNALERROR>   File "/home/yangliu/WAM2layers/WAM2layers/src/wam2layers/analysis/checks.py", line 13, in _warning_on_one_line
INTERNALERROR>     short_path = Path(filename).relative_to(Path(filename).parents[2])
INTERNALERROR>   File "/usr/lib/python3.8/pathlib.py", line 621, in __getitem__
INTERNALERROR>     raise IndexError(idx)
INTERNALERROR> IndexError: 2
```

I think the this function was created to make the warning message a bit shorter. However, it triggers the error above when running the test. Let me know if there is a better solution, other than changing the depth.

File: tests/test_data/README.md [hunk 2]
```diff
@@ -0,0 +1,12 @@
+### About how test data is prepared
+
+The [test data](./era5/) is already included in this directory, which can be used to run the [tests](../test_workflow.py).
+
+It is prepared through the following steps:
+- Download the required ERA5 fields from CDS <br>
+For model levels the following variables are needed:
+  - u, v, q on selected levels
+  - tp, e, sp, and tcw at the surface
+(see the example downloading script for more details about CDS qurey of ERA5 fields in `scripts/download_era5_ml.py`. Note that in the tests we use hourly data of corresponding ERA5 fields of the date 2022-08-31.)
```
Comment (Peter9192) [inline tests/test_data/README.md:10]:
```suggestion
(see the example downloading script for more details about CDS query of ERA5 fields in `scripts/download_era5_ml.py`. Note that in the tests we use hourly data of corresponding ERA5 fields of the date 2022-08-31.)
```

File: tests/test_data/README.md
Comment (Peter9192) [inline tests/test_data/README.md:1]: Should we provide attribution when redistributing the ERA5 data sample?

File: tests/test_workflow.py
Comment (Peter9192) [inline tests/test_workflow.py:1]: Would be nice to also test `wam2layers visualize output`, perhaps in a follow-up PR?

File: src/wam2layers/analysis/checks.py (hunk 1 above)
Comment (Peter9192) [reply src/wam2layers/analysis/checks.py:13]: Eventually I think we need to use these checks in another way, but I guess this is OK for now.

File: .github/workflows/ci.yaml [hunk 3]
```diff
@@ -23,17 +23,5 @@ jobs:
...
-      - if: ${{ steps.cache-sample-data.outputs.cache-hit != 'true' }}
-        name: Download sample data if not available in cache
-        run: |
-          gdown -O ~/wam2layers --folder https://rebrand.ly/wam2layers-example-data
-      - name: Run test case
-        run: |
-          cd ~/wam2layers/
-          wam2layers backtrack floodcase_202107.yaml
+      - name: Test workflow
+        run: pytest tests/test_workflow.py
```
Comment (Peter9192) [inline .github/workflows/ci.yaml:27]: Can you add a TODO here to run all the tests instead of just this one?

File: tests/test_workflow.py
Comment (geek-yang) [reply tests/test_workflow.py:1]: Sounds good. Let's do it in a follow-up PR to check the images output.

File: .github/workflows/ci.yaml (hunk 3 above)
Comment (geek-yang) [reply .github/workflows/ci.yaml:27]: Added.

File: tests/test_data/README.md (hunk 2 above)
Comment (geek-yang) [reply tests/test_data/README.md:10]: Well spot! Thanks.

File: tests/test_config.yaml
[the remark this answers is not in the document]
Comment (geek-yang) [reply tests/test_config.yaml:1]: Yes, agree. It is moved to `test_folder` now.

File: tests/test_data/README.md
Comment (geek-yang) [reply tests/test_data/README.md:1]: Good point! The following licence statement is now added to the data sample:
"Generated using Copernicus Climate Change Service information [2023]."

And in readme the following statement is added:
"License
Test data included here is generated using Copernicus Climate Change Service information and for more information about licensing, please check the [Licence Agreement](https://cds.climate.copernicus.eu/cdsapp/#!/terms/licence-to-use-copernicus-products) for Copernicus Products."

**Comment by geek-yang** [comment]
> Nice work! I left a few comments. Additionally, could you run black & isort and add a changelog entry?

Thanks for the review @Peter9192. The code is formatted with black and isort and an entry has been added to the changelog now. 😄

**Review by Peter9192 — APPROVED** [review]
Looks good, thanks!

**Pull request merged**